src/Catalyst.Core.Modules.Cryptography.BulletProofs/NativeBinding.cs

```diff
@@ -0,0 +1,219 @@
+            }
+
+            var signature = new byte[SignatureLength];
+            var public_key = new byte[PublicKeyLength];
+
+            var errorCode = std_sign(signature, public_key, privateKey, message, message.Length, context, context.Length);
+            if (errorCode != 0)
+            {
+                Error.ThrowErrorFromErrorCode(errorCode.ToString());
+            }
+
+            return new Signature(signature, public_key);
+        }
+        
+        [DllImport(Library, CallingConvention = CallingConvention.Cdecl)]
+        private static extern int std_verify(byte[] signature, byte[] publicKey, byte[] message, int messageLength, byte[] context, int contextLength, byte[] b);
+
+        internal static bool StdVerify(byte[] signature, byte[] publicKey, byte[] message, byte[] context)
+        {
+            if (signature.Length != SignatureLength)
+            { 
+                Error.ThrowArgumentExceptionSignatureLength(SignatureLength);
+            } 
+            
+            if (publicKey.Length != PublicKeyLength)
+            { 
+                Error.ThrowArgumentExceptionPublicKeyLength(PublicKeyLength);
+            } 
+
+            var isVerified = new byte[1];
+
+            var errorCode = std_verify(signature, publicKey, message, message.Length, context, context.Length,
+                isVerified);
+
+            if (errorCode != 0)
+            {
+                Error.ThrowErrorFromErrorCode(errorCode.ToString());
+            }
+
+            return isVerified[0] == 1;
+        }
+
+        [DllImport(Library, CallingConvention = CallingConvention.Cdecl)]
+        private static extern int validate_public_key(byte[] publicKey);
+
+        internal static void ValidatePublicKeyOrThrow(byte[] publicKey)
+        {
+            if (publicKey.Length != PublicKeyLength)
+            { 
+                Error.ThrowArgumentExceptionPublicKeyLength(PublicKeyLength);
+            } 
+
+            var errorCode = validate_public_key(publicKey);
+
+            if (errorCode != 0)
+            {
+                Error.ThrowErrorFromErrorCode(errorCode.ToString());
+            }
+        }
+
+        [DllImport(Library, CallingConvention = CallingConvention.Cdecl)]
+        private static extern void publickey_from_private(byte[] publicKey, byte[] privateKey);
+
+        internal static IPublicKey GetPublicKeyFromPrivate(byte[] privateKey)
+        {
+            if (privateKey.Length != PrivateKeyLength)
+            {
+                Error.ThrowArgumentExceptionPrivateKeyLength(PrivateKeyLength);
+            }
+
+            var publicKeyBytes = new byte[PublicKeyLength];
+            publickey_from_private(publicKeyBytes, privateKey);
+            
+            return new PublicKey(publicKeyBytes);
+        }
+
+        internal static string GetLastError()
+        {
+            var errorLength = last_error_length();
+            if (errorLength <= 0) return "";
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)